File: Leetcode/Leetcode-DSA-course/Linked-List/19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function removeNthFromEnd(head: ListNode | null, n: number): ListNode | null {
    let slow = head, fast = head, prev = null;
    for (let i = 0; i < n; i++) {
        fast = fast.next;
    }
    while (fast != null) {
        prev = slow;
        slow = slow.next;
        fast = fast.next;
    }
    if (prev == null) {
        head = head.next
        return head
    }
    prev.next = prev.next.next
    return head
};

File: Leetcode/Leetcode-DSA-course/Trees-and-Graphs/19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        slow = head
        fast = head
        prev = None
        for i in range(n):
            fast = fast.next
        while fast:
            prev = slow
            slow = slow.next
            fast = fast.next
        if prev == None:
            head = head.next
            return head
        prev.next = prev.next.next
        return head
